### ISSUE TITLE
Kyber patent license?

### ISSUE THREAD

**Issue by JamesTheAwesomeDude** [signalapp/libsignal#559]
Quick question: when you [chose Kyber](https://signal.org/blog/pqxdh/#:~:text=We%20believe%20that,foundations&text=already%20supported%20in%20the%20latest%20versions) for the post-quantum upgrade, did you pay any mind to U.S. patents 9094189 or 9246675?

NIST [says](https://csrc.nist.gov/csrc/media/Projects/post-quantum-cryptography/documents/selected-algos-2022/nist-pqc-license-summary-and-excerpts.pdf) that it already paid off the patentholders so everyone can use Kyber for free, but AFAICT **no member of the public has ever seen that license**. \(I'm not alone in thinking this, either; everyone from [Cisco](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/G0DoD7lkGPk/m/d7Zw0qhGBwAJ) to [MITRE](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/G0DoD7lkGPk/m/G_evJqQNKQAJ) to [the IETF](https://datatracker.ietf.org/meeting/116/proceedings#pquip) to [multiple](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/G0DoD7lkGPk/m/C1HORaCGAgAJ) [independent](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/G0DoD7lkGPk/m/of2mVEnlEAAJ) security researchers seems to agree both that the license is nowhere to be found, and that that's concerning.\)

I'm asking this question from 2 perspectives:

1. As a day-to-day Signal user in the United States, looking for proof that it's **legal** for me to use version 6.23.0 and newer \([`e2ef8e2`](https://github.com/signalapp/Signal-Android/commit/e2ef8e2ef9215dd058978cbdfac2d830b203ae9f)\) of Signal for Android.
2. As a FOSS developer [also struggling](https://github.com/JamesTheAwesomeDude/pypqc/issues/8) with this same licensing ambiguity, curious how other major FOSS projects have handled it.

**Comment by ehrenkret-signal** [comment]
In choosing the CRYSTALS-KYBER algorithm, we relied on NIST's statement affirming that it was permissible to use. See: https://csrc.nist.gov/csrc/media/Projects/post-quantum-cryptography/documents/selected-algos-2022/nist-pqc-license-summary-and-excerpts.pdf

**Comment by JamesTheAwesomeDude** [comment]
Indeed, I've already read that strange press release in its entirety. \(It was the *first link* in my question, after Signal's own blog post.\) That PDF hasn't been signed by the patentholders or by NIST's legal team, and it doesn't elaborate on any secret "poison pills" which might revoke anyone's permission to use Kyber.

Per Bernstein's finding, linked in my original question, this is **not a theoretical worry**; there is actual precedent for this happening. And NIST has [left this intentionally ambiguous](https://mailarchive.ietf.org/arch/msg/pqc/ArqjlUc2cVmRQZeavDRDoXHwkyo/#:~:text=can't%20promise%20anything) in the case of Kyber.

-----

As a concrete example: if I am a party to a lawsuit with Algo Consulting for an unrelated reason, it's possible I would **lose my license to operate** Signal versions 6.23 and newer, but the incomplete EULA Signal's currently displaying [doesn't *warn* us](https://github.com/signalapp/Signal-Android/blob/9071fd0024954ade4f27e3e72604b2023f15b0d5/LICENSE#L607) about this.

**Comment by JamesTheAwesomeDude** [comment]
At the very least, could you consider including some *mention* of the `Kyber License` on the Third-Party Licenses page?

https://github.com/signalapp/Signal-Android/blob/9071fd0024954ade4f27e3e72604b2023f15b0d5/app/src/main/res/raw/third_party_licenses

I know that this is a very awkward ask, both because the license isn't OSI-approved and because the full text of the license isn't even available, but that awkwardness only seems like all the \_more\_ reason to include it on the page that's \_supposed\_ to inform the user of any T&C beyond the main codebase's license.

**Comment by ehrenkret-signal** [comment]
We will add a link to the NIST PDF for the Kyber Patent License to https://github.com/signalapp/libsignal/tree/main/acknowledgments and respective client platform specific equivalents.

**Comment by jrose-signal** [comment]
I think we're sticking with the link for now (added to the acknowledgments files in 0.41.0, will eventually make it into apps). Those screens aren't really an ideal viewing experience regardless, but that's a larger issue than just how we as an organization consider our use of Kyber to be patent-abiding. But thank you for bringing it up!

**Comment by JamesTheAwesomeDude** [comment]
I am aware that you "consider" your use of Kyber to be "patent-abiding", but it's absolutely inappropriate to conceal the full license from the users &mdash; especially considering that, as I noted earlier, the AGPL includes a **limitation of warranty of noninfringement**, meaninng users are *not* legally protected by your "consideration".

You include the full text of the Apache and MIT and other 3rd-party licenses that may affect Signal. So why make the decision to *exclude* the full text of the Kyber license, which might actually impact users, instead hiding it behind a link that **can't even be clicked** in the app?
